app/models/segmentation_statistic.rb

```diff
@@ -0,0 +1,602 @@
+      data: generate_monthly_segmentation_metrics( at_time ),
+      created_at: at_time.beginning_of_day
+    )
+  end
+
+  def self.stats_generated_for_day?( at_time = Time.now )
+    SegmentationStatistic.where( "DATE(created_at) = DATE(?)", at_time.utc ).exists?
+  end
+
+  def self.get_viz_stats_for_day( at_time = Time.now )
+    stats = SegmentationStatistic.where( "DATE(created_at) = DATE(?)", at_time.utc )
+    format_for_viz( stats.first.data ) if stats.exists?
+  end
+
+  # private
+
+  # Generate monthly metrics
+  def self.generate_monthly_segmentation_metrics( at_time = Time.now )
+    monthly_segmentation_data = generate_monthly_segmentation_data( at_time )
+    generate_main_metrics( monthly_segmentation_data )
+  end
+
+  # Generate all the users data for a month
+  def self.generate_monthly_segmentation_data( at_time = Time.now )
+    monthly_segmentation_data = {}
+    day = at_time
+    30.times do
+      users_from_obs_data( day, monthly_segmentation_data )
+      users_from_ids_data( day, monthly_segmentation_data )
+      users_from_kibana_data( day, monthly_segmentation_data )
+      day -= 1.days
+    end
+    users_from_db( at_time, monthly_segmentation_data )
+    monthly_segmentation_data
+  end
+
+  # Generate all the users data for a single day
+  def self.generate_daily_segmentation_data( at_time = Time.now )
+    daily_segmentation_data = {}
+    day = at_time
+    users_from_obs_data( day, daily_segmentation_data )
+    users_from_ids_data( day, daily_segmentation_data )
+    users_from_kibana_data( day, daily_segmentation_data )
+    users_from_db( day, daily_segmentation_data )
+    daily_segmentation_data
+  end
+
+  # Default user structure
+  def self.default_user( user_id )
+    {
+      id: user_id,
+      active: false,
+      new: false,
+      created_at: -1,
+      total_obs: 0,
+      total_ids: 0,
+      obs: 0,
+      ids: 0,
+      web_obs: 0,
+      ios_obs: 0,
+      android_obs: 0,
+      seek_obs: 0,
+      other_obs: 0,
+      q: 0,
+      web_q: 0,
+      ios_q: 0,
+      android_q: 0,
+      seek_q: 0,
+      other_q: 0
+    }
+  end
+
+  # Get users having submitted observations
+  # including the number of observations by application (android, ios, seek, web) for each
+  def self.users_from_obs_data( day, users_data )
+    filter = [
+      {
+        range: {
+          created_at: {
+            gte: day.beginning_of_day,
+            lte: day.end_of_day
+          }
+        }
+      }
+    ]
+    agg = {
+      user_id: {
+        terms: {
+          field: "user.id",
+          size: 10_000_000
+        },
+        aggs: {
+          app_id: {
+            terms: {
+              field: "oauth_application_id",
+              size: 10_000_000
+            }
+          }
+        }
+      }
+    }
+
+    obs_data = Observation.elastic_search(
+      size: 0,
+      filters: filter,
+      aggregate: agg
+    ).response.aggregations.user_id.buckets
+
+    obs_data.each do | u |
```

Comment:
I find code more readable when variable names are more descriptive. Instead of `u` I might have used `user_counts_bucket` or just `bucket` since that's what it is - one bucket from an Elasticsearch aggregation response. Generally I prefer very descriptive variable names where its pretty clear what they are immediately without having to look back, even if that means very long variable names